### ISSUE TITLE
Add an AMP argument type for passing a file descriptor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5650]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5650|
|Type|enhancement|
|Created|2012-04-20 21:45:12Z|
|Branch|https://github.com/twisted/twisted/tree/amp-descriptor-5650-2|

[#5615](https://github.com/twisted/twisted/issues/5615) adds support to UNIX transports for sending file descriptors between processes.  We should add an argument type to AMP that makes it trivial to use this functionality over an AMP connection.

A new argument type, `Descriptor`, would take a file descriptor, use `sendFileDescriptor` and such to get it into the process on the other side, and hand it back to application code as the value for the argument (or response).

<details><summary>Searchable metadata</summary>

```
trac-id__5650 5650
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_amp_descriptor_5650_2 branches/amp-descriptor-5650-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__amp amp
time__1334958312000000 1334958312000000
changetime__1336161261000000 1336161261000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [34244]) Branching to 'amp-descriptor-5650'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [34291]) Branching to 'amp-descriptor-5650-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Added one.  Pretty straight forward.  Relying strictly on ordering seems maybe not ideal, but it will work in all the cases I can think of, at least.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/amp-descriptor-5650-2)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set owner to @exarkun|
|-|-|

Nice feature, having file descriptor transfer be so easy is very useful.

  1. I'd want the roundtrip test to use two protocol instances, not just the one. 
  1. Implementation details would be clearer with maybe one extra sentence in the `Descriptor` docstring, some better written version of "Because file descriptors are transmitted out-of-band, and the fd number on the receiving side may be different than on the sending side, counting the number of transmitted fds is a good way to refer to them within the AMP protocol."
  1. One of the lines of Descriptor docstring is more than 80 characters.
  1. `Descriptor.fromStringProto` is breaking abstraction boundaries; add a method to `_DescriptorExchanger` for retrieving fds.
  1. Need tests for the `_DescriptorExchanger` methods. In particular, I'd want validation that transmitting and receiving multiple fds worked correctly, not just the one implied by the roundtrip test, but also that file descriptors are actually transmitted.
  1. What happens if you send a `Descriptor` over a transport that can't send file descriptors? I suspect the error would be AttributeError, something better would be nice.
  1. Might want to talk about `bytes` rather than `strings` in prose, for future compatibility.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.

> 1. I'd want the roundtrip test to use two protocol instances, not just the one.

Fixed in r34347

> 1. Implementation details would be clearer with maybe one extra sentence in the Descriptor docstring, some better written version of "Because file descriptors are transmitted out-of-band, and the fd number on the receiving side may be different than on the sending side, counting the number of transmitted fds is a good way to refer to them within the AMP protocol."

Tried to improve this in r34348.  For some reason, I have a lot of difficulty with all of the language related to relying on the ordering.  I'm missing a word that means "a continuous, monotonically increasing integer sequence starting at zero in which each element of the sequence is assigned to a Descriptor argument usage according to the order in which those arguments are encoded, with each Descriptor that is encoded being associated with the next unused element from the sequence."  So I feel a lot of the documentation trying to explain that is rather clumsy.

> 3. One of the lines of Descriptor docstring is more than 80 characters.

I think two are.  They're both lines that are almost all taken up by a `L{}`, which I'd rather not break across multiple lines, since that requires inserting some possibly-incorrect whitespace.

> 4. Descriptor.fromStringProto is breaking abstraction boundaries; add a method to _DescriptorExchanger for retrieving fds.

Hmmmm.  I'm not sure I agree.  `Descriptor` and `_DescriptorExchanger` are the two halves of the implementation of this feature.  Still, I made the code use a private method instead of a private attribute in r34349.

> 5. Need tests for the _DescriptorExchanger methods. In particular, I'd want validation that transmitting and receiving multiple fds worked correctly, not just the one implied by the roundtrip test, but also that file descriptors are actually transmitted.

I don't understand what test you're asking for here.

> 6. What happens if you send a Descriptor over a transport that can't send file descriptors? I suspect the error would be AttributeError, something better would be nice.

There's no defined way for an AMP argument type to signal failure.  Perhaps there should be, but that seems like a more general AMP improvement to me, not a blocker for this ticket.

> 7. Might want to talk about bytes rather than strings in prose, for future compatibility.

Done in r34349.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set owner to @exarkun|
|-|-|

OK, I think I'm happy with the current state (I re-read tests and realized you actually were covering what I thought hadn't been). Feel free to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [34350]) Merge amp-descriptor-5650-2

Author: exarkun
Reviewer: itamar
Fixes: [#5650](https://github.com/twisted/twisted/issues/5650)

Add a new AMP argument type, `Descriptor`, for sending file descriptors over
UNIX transports.
